_format/1.1/index.md

```diff
@@ -1839,3 +1877,572 @@ An error object **MAY** have the following members:
 [error details]: #errors
 [member names]: #document-member-names
 [pagination]: #fetching-pagination
+
+## <a href="#operations" id="operations" class="headerlink"></a> Operations
+
+A server **MAY** support usage of the `operations` array to complete one or more
+fetch or mutation operations in a single request.
+
+All HTTP requests that include an `operations` array **MUST** be issued with
+`PATCH`.
+
+For the sake of clarity, requests that include operations are called
+"operations requests" while requests that perform a single fetch or mutation are
+called "singular requests".
+
+### <a href="#operation-objects" id="operation-objects" class="headerlink"></a> Operation Objects
+
+An operation object **MAY** contain any of the following members:
+
+* `op`: an operation code, expressed as a string, that indicates the type of
+  operation to perform. Valid values include:
+  - `"get"`: fetches a resource, resource identity, collection of resources, or
+     collection of resource identitities
+  - `"add"`: creates a new resource or relationship
+  - `"update"`: updates a resource or relationship
+  - `"remove"`: removes a resource or relationship
+
+* `data`: the operation's "primary data".
+
+* `included`: an array of [resource
+  objects](#document-resource-objects) that are
+  related to the operation's primary data and/or each other
+  ("included resources").
+
+* `ref`: the resource or resource collection that is the target of the
+  operation. This member serves to disambiguate the target at a particular
+  endpoint if that endpoint represents multiple resources. `ref` **MUST** be
+  expressed as an object that **MAY** contain any of the following members:
+  - `type`: resource `type`
+  - `id` or `lid`: resource identifier or local identifier
+  - `relationship`: the name of the relationship
+
+* `params`: all of the query parameters allowed for singular requests can be
+  represented in JSON form in an operation's `params` object. Possible members
+  include:
+  - `include`
+  - `sort`
+  - `filter`
+  - `page`
+  - `fields`
+
+* `links`: a [links object](#document-links) related to the operation's primary
+  data.
+
+* `meta`: a [meta object](#document-meta) that contains non-standard
+  meta-information about the operation.
+
+Different members are required for processing different types of operations, as
+described below.
+
+> Note: `op` codes are used instead of the HTTP methods prescribed for singular
+  requests. The values `"get"`, `"add"`, `"update"`, and `"remove"` align with
+  the HTTP methods `GET`, `POST`, `PATCH`, and `DELETE` used for singular
+  requests.
+
+### <a href="#processing-operations" id="processing-operations" class="headerlink"></a> Processing Operations
+
+A server **MUST** perform operations in the order they appear in the
```

Comment:
I keep going back and forth on this. On the one hand, a server can't possibly know the dependencies of every operation. On the other, it seems like it exposes too much of the server implementation to the client.

Take the example of a client which wants to replace a resource. If the operations happen out of order, then the server will generate a collision error (when the _add_ happens before the _remove_).

If the order of operations given by the client must be followed, then it's the client's responsibility to order these correctly. That makes sense because it would be far too difficult for a server to infer the correct order in every possible scenario.

However, take the example of creating a resource which has a relationship to another resource. There are 3 possible orderings that could be valid or invalid dependent on the storage implementation of the server. (1) If the server stores relationships separately from resources, then both resources need to be saved before the relationship is created. (2) If the server stores relationships on the referencing resource, then the the referenced resource needs to be created first. (3) Finally, if the relationship is stored on the _referenced_ resource, then the opposite is true and the _referencing_ resource must be saved first. When the responsibility is on the client to know this, the order of operations needs to change based on the implementation details of the server. This seems incorrect and difficult to manage, generic clients will need to switch behavior based on something not defined in the spec.

After writing that out, I wonder if the way to handle this is with an opt-in dependency tree. That is, allow operations to be identified and also allow operations to specify operations on which they depend. When operations have no dependency or when multiple operations depend on the same operation, the server may perform those operations in an arbitrary order. That would allow a server to make a best-effort at ordering the operations appropriately (e.g. it could order operations according to relationships) but it would also allow the client to express order when it would otherwise be too ambiguous for the server to determine.

Reply:
Here's an example of the replace example from above with dependencies on other operations.

```json
PATCH /bulk
{
  "operations": [{
    "op": "add",
    "deps": ["remove_some_resource"],
    "data": {"not": "shown"}},
  {
    "op": "remove",
    "id": "remove_some_resource",
    "ref": {"not": "shown"}}
  ]
}
```